internal/services/sources_service.go

```diff
@@ -47,3 +49,39 @@ func GetSource(w http.ResponseWriter, r *http.Request) {
 		renderError(w, r, payloads.NewRenderError(r.Context(), "show source", err))
 	}
 }
+
+func fetchARN(ctx context.Context, client sources.SourcesIntegration, sourceId string) (string, error) {
+	// Get all the authentications linked to a specific source
+	resp, err := client.ListSourceAuthenticationsWithResponse(ctx, sourceId, &sources.ListSourceAuthenticationsParams{}, AddIdentityHeader)
+	if err != nil {
+		return "", fmt.Errorf("cannot list source authentication: %w", err)
+	}
+
+	// Filter authentications to include only auth where resource_type == "Application"
+	// TODO: use sources client filtering after they solve their bug
+	auth := filterSourceAuthentications(resp.JSON200.Data)
+	// Get the resource_id which equals to application_id
+	// and check that application_type_id in /applications/<app_id> equals to provisioning id
+	res, err := client.ShowApplicationWithResponse(ctx, *auth.ResourceId, AddIdentityHeader)
+	if err != nil {
+		return "", fmt.Errorf("cannot list source authentication: %w", err)
+	}
+
+	if *res.JSON200.ApplicationTypeId == config.Sources.AppTypeId {
+		return *auth.Username, nil
+
+	}
+	return "", fmt.Errorf("cannot find authentication linked to source id %s and to the provisioning app: %w", sourceId, err)
+}
+
+func filterSourceAuthentications(authentications *[]sources.AuthenticationRead) sources.AuthenticationRead {
+	auths := *authentications
+	list := make([]sources.AuthenticationRead, 0, len(auths))
+	for _, auth := range auths {
+		if *auth.ResourceType == "Application" {
+			list = append(list, auth)
+		}
+	}
+	//TODO: assumption: each source has one authentication linked to
```

Comment:
A left over? This needs to be implemented.

This function must return error when it would return invalid data.

Reply:
Is that enough? :)
```golang
// Assumption: each source has one authentication linked to it
	if len(list) > 1 {
		return sources.AuthenticationRead{}, fmt.Errorf("there is more then one authentication linked to this source %s", sourceId)
	}
	return list[0], nil
```